lib/base/configuration.cpp

```diff
@@ -3,12 +3,23 @@
 #include "base/configuration.hpp"
 #include "base/configuration-ti.cpp"
 #include "base/exception.hpp"
+#include <boost/asio.hpp>
 
 using namespace icinga;
 
 REGISTER_TYPE(Configuration);
 
-String Configuration::ApiBindHost{"::"};
+String Configuration::ApiBindHost = []() {
+	try {
+		// Auto-detect if IPv6/AF_INET6 is supported by the system by trying to open a socket.
+		boost::asio::io_context io_context;
+		boost::asio::ip::tcp::socket sock(io_context, boost::asio::ip::tcp::v6());
+	} catch (const std::exception& e) {
```

Comment:
I was considering doing that at first and decided against that because this would probably need more special handling for Windows and even then, what do you do if it's another error? Assume IPv6 is (not) supported? Consider it a fatal error?

Reply:
1. _Probably_? "Do or do not. There is no try." – Yoda
2. Assume... ah, I get what you mean... and `getaddrinfo(3)` (or Boost equivalent) w/ AI_NUMERICHOST?

Reply:
Apropos _fatal error_: I’d put `boost::asio::io_context io_context;` out of `try{}`.

Reply:
> 2. Assume... ah, I get what you mean... and `getaddrinfo(3)` (or Boost equivalent) w/ AI_NUMERICHOST?

Looks like `getaddrinfo` is of little help here:

```pycon
>>> from socket import *; from pprint import *
>>> pprint(getaddrinfo("::", 5665, family=AF_INET6, type=SOCK_STREAM, flags=AI_PASSIVE|AI_NUMERICHOST))
[(<AddressFamily.AF_INET6: 10>,
  <SocketKind.SOCK_STREAM: 1>,
  6,
  '',
  ('::', 5665, 0, 0))]
>>> pprint(getaddrinfo("::", 5665, family=AF_UNSPEC, type=SOCK_STREAM, flags=AI_PASSIVE|AI_NUMERICHOST))
[(<AddressFamily.AF_INET6: 10>,
  <SocketKind.SOCK_STREAM: 1>,
  6,
  '',
  ('::', 5665, 0, 0))]
>>> socket(AF_INET6)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/socket.py", line 151, in __init__
    _socket.socket.__init__(self, family, type, proto, fileno)
OSError: [Errno 97] Address family not supported by protocol

```

Reply:
Now as you have a test sys for this: what happens if you try to provoke a non-EAFNOSUPPORT error w/ AF_INET6?

Reply:
With the current state of the PR? It would assume that IPv6 is not supported and later attempt to bind to IPv4 only. But I think this will then likely fail with the same error. The only easy way to cause another error would be `EMFILE` by restricting the number of file descriptors using ulimit.

Reply:
1. No, with Python.
2. I’ve got something easier:

```
>>> socket(AF_INET,SOCK_STREAM,42)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/socket.py", line 232, in __init__
    _socket.socket.__init__(self, family, type, proto, fileno)
OSError: [Errno 43] Protocol not supported
>>> socket(AF_INET,SOCK_STREAM,0)
<socket.socket fd=3, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('0.0.0.0', 0)>
>>>
```

Reply:
And what do we now do with this information?

Reply:
The one **I posted**? Nothing. I'd like **you** to do this on your test sys w/ s/AF_INET/AF_INET6/.

Reply:
```pycon
>>> socket(AF_INET,SOCK_STREAM,42)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/socket.py", line 151, in __init__
    _socket.socket.__init__(self, family, type, proto, fileno)
OSError: [Errno 93] Protocol not supported

>>> socket(AF_INET,SOCK_STREAM,0)
<socket.socket fd=4, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('0.0.0.0', 0)>

>>> socket(AF_INET6,SOCK_STREAM,42)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/socket.py", line 151, in __init__
    _socket.socket.__init__(self, family, type, proto, fileno)
OSError: [Errno 97] Address family not supported by protocol

>>> socket(AF_INET6,SOCK_STREAM,0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/socket.py", line 151, in __init__
    _socket.socket.__init__(self, family, type, proto, fileno)
OSError: [Errno 97] Address family not supported by protocol
```

So in this case, the address family is checked first. But that sounds like an implementation detail I would not rely on too much. (But in our case, the other parameters should be valid in the end. We want to open a TCP socket later anyway, so if TCP wasn't supported for whatever reason, good luck getting Icinga to run.)

Reply:
I consider it a reliable implementation detail. According `socket(2)` there are errors of the following categories:

1. low on resources
2. invalid input
3. access denial

IMAO a good implementation would always check the input, at least before trying to allocate resources. Same for access. And all of the input is valid. Even if TCP were disabled, you’d get another errno – fine!